### ISSUE TITLE
When a Draw interaction of vector layer is in use, all tile and image layers are also being rerender

### ISSUE THREAD

**Issue by Feofilakt** [openlayers/openlayers#15722]
**Describe the bug**
Don't know if it is a bug or intended behavior. It seems that when a Draw interaction of vector layer is in use, all tile and image layers are also being rerender. We have faced performance problems because of it (high GPU usage during simple interaction). Maybe I'm missing something here - unfortunatly, am not an expert in OL rendering system.

**To Reproduce**
1. Go to https://codepen.io/Feofilakt/pen/mdgLzPr
You can see captions "Hello world 0" and "Hello world 1". Each on one of canvases - I checked it by removing single canvas:

![image](https://github.com/openlayers/openlayers/assets/5498396/b9a3bb2c-7fbc-4c01-be50-6a877aa1e490)

![image](https://github.com/openlayers/openlayers/assets/5498396/f5521871-50d5-47de-a731-527314dd29c1)

2. Try to draw line (click and move cursor)
Both captions disappear.

**Expected behavior**
Only one caption disappears, other (on the canvas of tile layer) still remains.

**Comment by ahocevar** [comment]
This is intended behavior - for any change anywhere on the map, the whole map is re-rendered. Usually what's being done is just the composition of the final map canvas, so that should not cause any performance problems.

It would be interesting to hear in which scenario you're facing performance problems. I guess not a simple one like in the code pen you have prepared?

**Comment by ahocevar** [comment]
Make sure all image and tile layers have the same resolution, and avoid setting `className` on layers, so they can be combined onto a single `canvas`. But hard to give more detail without seeing actual code.

**Comment by Feofilakt** [comment]
Its just too much rendering, I think: all the layers on big map. Not a problem by itself, but it happens in real time (on each mouse move event?) and so interfere with interaction sketch rendering (which, I guess, has same priority).

**Comment by ahocevar** [comment]
The rendering engine is built for that - when you pan the map, it also re-renders on every pointer move. Check your code for the pitfalls I mentioned above. In the case of image layers, it might also be the case that an image cannot be converted to an ImageBitmap, which is required for fast composition. You can check for that with a breakpoint in `ol/Image.js`'s `decode()` function and see if it runs into one of the fallbacks.

**Comment by Feofilakt** [comment]
You were right about ImageBitmap. We use custom loader for POST requests and function actually runs into fallback. After changing that performance have imroved a little, but unfortunatly lags are still noticable.

**Comment by ahocevar** [comment]
Are all your images now being converted to `ImageBitmap`s? If not, make them smaller (e.g. by setting `ratio: 1` in the source or loader).

**Comment by Feofilakt** [comment]
Yes, all images are now being converted to ImageBitmap and ratio is 1. Probably we will be forced to implement draw without sketch.

**Comment by ahocevar** [comment]
Did you also remove all className configs from the layers?

**Comment by Feofilakt** [comment]
Yes, we did that.

**Comment by ahocevar** [comment]
@Feofilakt If you can provide a live example that shows the problem, we could profile it and see what the culprit is.

**Comment by alexerisov** [comment]
Hello. I faced with similar issue and had an idea, maybe make a fork and make a correction so that skip render of other layers on each "pointermove” event while drawing a sketch ? Could these fixes break something?

**Comment by ahocevar** [comment]
This would be quite a hack. And you'd have to put the draw layer on a separate canvas. It's better/easier to find out why you're spending so much time in the composition phase. This is why I'd like to see this issue in a live example I can profile.

**Comment by alexerisov** [comment]
The issue here is not with the composition time, but with the GPU load. In other words, it’s not a library bug; it’s simply that the GPU is struggling. I’ve observed a similar issue even in online paint. Providing a demo is challenging, but I believe you could achieve this effect by heavily loading the GPU (e.g., using a benchmark) and then using the Draw interraction.

Regarding the separate canvas, yes, that was my initial attempt. I added className: "draw-layer" [here](https://github.com/openlayers/openlayers/blob/a84aac17c71cc5707e9a2bf33639d91ccdf79207/src/ol/interaction/Draw.js#L888). However, the main canvas still updates during events. Currently, the idea is to add [here](https://github.com/openlayers/openlayers/blob/a84aac17c71cc5707e9a2bf33639d91ccdf79207/src/ol/MapBrowserEventHandler.js#L315) a simple check (smth like (`event.originalEvent.target.className === 'draw-layer`) и  ('layer !== sketchLayer`) and skip and ignore events for all layers except sketchlayer.

**Comment by ahocevar** [comment]
@alexerisov It's not the pointermove event that triggers map rendering, so your idea won't work. What would work is in the map's Composite renderer's `renderFrame()` method  if all layers that can be grouped in the same container (`useContainer()`) did not change since the previous render frame. If they did not, the previous render frame's canvas can be used again.

**Comment by alexerisov** [comment]
Thanks for the help, I'll try your  suggestion  :)

**Comment by ahocevar** [comment]
@alexerisov It's a corner case, and would require quite a lot of code for this corner case. The only situation where rendering would get faster would be drawing with a sketch that changes on pointermove. So my suggestion would be to set a style function that returns undefined on the Draw interaction.

**Comment by alexerisov** [comment]
Yes, I thought about similar solution but it intercepts 'pointermove' at all, so sketch layer isn't updating (similar to your first suggestion with `style: () => undefined`.
Maybe you have any idea how to intercept 'pointermove' but keep skecth layer updating?

**Comment by ahocevar** [comment]
Sketch layer updates means re-compose the map image, so you can either have one or the other.

**Comment by alexerisov** [comment]
Hm, I moved sketch layer to another canvas (by adding `className` to Draw.overlay_). Should map re-compose this case?

p.s. Also, regarding the editing of the rendering-logic, is it possible to enforce such a change? For instance, on a commercial basis ? If possible, what about time/cost? We are interested in such optimization (and it would be beneficial for the community - prevent the layers from uneccessary renders.

**Comment by ahocevar** [comment]
> Hm, I moved sketch layer to another canvas (by adding className to Draw.overlay_). Should map re-compose this case?

Yes, the map will always be re-composed when anything on any layer changes.

> p.s. Also, regarding the editing of the rendering-logic, is it possible to enforce such change for openlayers? E.g on a commercial basis ? If possible, what about time/cost? We are interested in such optimization (and it would be beneficial for the community - prevent the layers from uneccessary renders.

Like I said, the change you're envisioning would add a lot of code and require the user to split the map into render sections, i.e. by setting `className`, which is something I'd like to avoid. I see the `className` as an option that is possible due to an implementation detail, but it would be better to remove it.

A more straightforward alternative I could think of would be to uncouple unmanaged layers (i.e. those you don't add to the map with `map.addLayer(layer)`, but `layer.setMap(map)`, which is what the `Draw` interaction uses, from the map renderer's main rendering loop. I'd be willing to make such a change if you sponsor me on https://github.com/sponsors/ahocevar/ with a monthly contribution of 200$ for at least 12 months.

**Comment by ahocevar** [comment]
> At the moment, Layer.render() calls from different cases and can't be controlled\overrided

Well, you can call `render()` on a layer any time you want, and if you have made sure the layer has its own separate canvas (by setting a distinct `className`), you can trigger a render any time. But that won't help you with your use case, because you want the opposite, i.e. to avoid that other layers' canvases get re-rendered. Which you cannot, because a change in a geometry of the draw layer's source triggers a re-render on the whole map.

>  It would be easier to define render-condition for each Layer

Yes, and that might be a sensible addition of the API, but is currently unintuitive because it requires a combination with the `className` setting.

So my idea was to make the most common use cases as intuitive as possible. And I think the most common use case is overlays with draw sketches or feature selections. For both, an unmanaged layer makes sense. And it is a clear distinction by which we can say: a layer/source change does not trigger a map render cycle, but only a render cycle of the unmanaged layers. The composite map renderer could measure the time spent in a whole render cycle, and render the unmanaged layers to a separate canvas more frame budget than a certain threshold was spent in the last render cycle. A clean solution with not too much additional code, and nothing the user needs to do to make it work.
